statsmodels/tsa/vector_ar/var_model.py

```diff
@@ -392,121 +593,168 @@ def fit(self, maxlags=None, method='ols', ic=None, trend='c',
+                 " given in the model constructor.", FutureWarning)
+
+            k_ar = maxlags
+        else:
+            try:
+                assert hasattr(self, "k_ar")
+            except:
+                raise ValueError("Please give nlags to the model constructor "
+                        "before calling fit.")
+            k_ar = self.k_ar
 
         if trend not in ['c', 'ct', 'ctt', 'nc']:
             raise ValueError("trend '{}' not supported for VAR".format(trend))
 
         if ic is not None:
-            selections = self.select_order(maxlags=maxlags, verbose=verbose)
+            warn("The ic argument to fit is deprecated. Please use the "
+                 "select_order classmethod or select_order_fit instead.",
+                 FutureWarning)
+            selections = self.select_order(self.Y, maxlags=maxlags,
+                                           verbose=verbose)
             if ic not in selections:
                 raise Exception("%s not recognized, must be among %s"
                                 % (ic, sorted(selections)))
-            lags = selections[ic]
+            k_ar = selections[ic]
             if verbose:
-                print('Using %d based on %s criterion' %  (lags, ic))
+                print('Using %d based on %s criterion' %  (k_ar, ic))
         else:
-            if lags is None:
-                lags = 1
+            if k_ar is None:
+                k_ar = 1
 
         k_trend = util.get_trendorder(trend)
-        self.exog_names = util.make_lag_names(self.endog_names, lags, k_trend)
-        self.nobs = len(self.endog) - lags
+        #TODO: anything we attach here might be stale on refit. Test.
+        self.nobs = len(self.Y) - k_ar
+        self.trend = trend
 
-        return self._estimate_var(lags, trend=trend)
+        return self._estimate_var(k_ar, trend=trend)
 
-    def _estimate_var(self, lags, offset=0, trend='c'):
+    #NOTE: maybe we should have a select_sample method in TSAModel to do what
+    #      offset is doing?
+    def _estimate_var(self, k_ar, offset=0, trend='c'):
         """
-        lags : int
-        offset : int
-            Periods to drop from beginning-- for order selection so it's an
-            apples-to-apples comparison
-        trend : string or None
-            As per above
+        The offset is an int. For the number of periods to drop from
+        beginning so that the sample size is comparable when doing order
+        selection.
         """
         # have to do this again because select_order doesn't call fit
         self.k_trend = k_trend = util.get_trendorder(trend)
 
         if offset < 0: # pragma: no cover
             raise ValueError('offset must be >= 0')
 
-        y = self.y[offset:]
+        Y = self.Y[offset:]
+
+        # lagged Y, includes trend variable(s) in first column(s)
+        X = util.get_lagged_y(Y, k_ar, trend=trend, has_constant='raise')
```

Comment:
possible temporary solution for backwards compatibility:

create `initialize` method that creates the regressor array, that can be called during deprecation without having the duplicate `get_lagged_endog` in several methods